File: Algorithms/data_structures/heap_unionfind/make_heap/make_heap.py
#python3
import sys

n = int(sys.stdin.readline())
array = list(map(int, sys.stdin.readline().split()))
def left_child(i):
	return i*2 + 1
def right_child(i):
	return i*2 + 2

def make_heap(n, array):
	operations = []
	lens = n//2

	def shift_down(index):
		while(True):
			minChild = index
			left = left_child(index)
			right = right_child(index)
			if (left < len(array) and array[left] < array[minChild]):
				minChild = left
			if (right < len(array) and array[right] < array[minChild]):
				minChild = right
			if(index != minChild):
				operations.append((index, minChild))
				array[index], array[minChild] = array[minChild], array[index]
				index = minChild
			else:
				break;

	while(lens >= 0):
		shift_down(lens)
		lens -= 1
	print(len(operations))
	for operation in operations:
		print(operation[0], operation[1])

make_heap(n, array)

File: Algorithms/data_structures/heap_unionfind/make_heap/make_heap.js
function parent(i) {return Math.floor((i - 1) / 2)};
function leftChild(i) {return i*2 + 1;};
function rightChild(i) {return i*2 + 2};

function makeHeap(n, array) {
  var results = '';
  var operations = 0;
  function shiftDown(index) {
    var left, right, minChild;
    while(true) {
      minChild = index;
      left = leftChild(index);
      right = rightChild(index);
      if(left < n && array[left] < array[minChild])
        minChild = left;
      if(right < n && array[right] < array[minChild]) 
        minChild = right;
      if(minChild != index) {
        results += index + ' ' + minChild + '\n';
        operations++;
        array[index] = [array[minChild], array[minChild] = array[index]][0];
        index = minChild;
      } else {
        break;
      }
    }
    
  }
  var limit = Math.floor(n/2);
  for (var i = limit; i >= 0; i--) {
    shiftDown(i);
  }
  process.stdout.write(operations + '\n' + results);
}

var readline = require('readline')
var rl = readline.createInterface({
      input: process.stdin,
      terminal: false});

rl.on('line', readLine);

var n = -1
var array = []
var lines = 0;

rl.on('close', function() {
  makeHeap(n, array);
  process.exit();
})

function readLine(line) {
  if(line != null) {
    if(++lines == 1){
      n = parseInt(line.toString());
    } else if(lines == 2) {
      array = line.toString().split(' ').map(function(num) {return parseInt(num)});
    }  
  }
  
}


